### PR TITLE
envoy: Build Istio Docker images

### PR DESCRIPTION
**Pull request by rlenglet**
Build Istio filters into Envoy binary.
Split Envoy binary build targets for Cilium and Istio.
Only build the Istio filters into the binary used in Istio sidecar
proxy containers.

Define Makefile targets to build Istio 0.8.0+ proxy images:
docker-istio-proxy builds image cilium/istio_proxy:*
docker-istio-proxy-debug builds image cilium/istio_proxy_debug:*

Add envoy/README.Istio.md to specify Istio image build process.

Signed-off-by: Romain Lenglet <romain@covalent.io>

### CODE REVIEW

**Comment by rlenglet** [comment]
test-me-please

**Comment by rlenglet** [comment]
CI build failure:
https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-Validated/2768/
Cilium agent doesn't seem to start:
```
 Get http:///var/run/cilium/cilium.sock/v1/healthz: dial unix /var/run/cilium/cilium.sock: connect: no such file or directory
```

File: envoy/Dockerfile.istio_proxy_debug.in [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+chmod 2755 /usr/local/bin/envoy
+
+# Override the Envoy bootstrap to configure Envoy to use API v2 and to define
+# the "xds-grpc-cilium" cluster used by the Cilium filters to connect to Cilium
+# agent via a Unix domain socket.
+# This file is adapted from Istio's API v2 bootstrap:
+# https://github.com/istio/istio/blob/67be0412218829febe6539f5360b3f2ce1716adb/tools/deb/envoy_bootstrap_v2.json
+COPY envoy_bootstrap_tmpl.json /var/lib/istio/envoy/envoy_bootstrap_tmpl.json
+
+ENTRYPOINT ["/usr/local/bin/pilot-agent"]
```
Comment (jrajahalme) [inline envoy/Dockerfile.istio_proxy_debug.in:22]: This file looks identical to `Dockerfile.istio_proxy.in`, do we need both?

File: envoy/Makefile [hunk 2]
```diff
@@ -40,28 +54,46 @@ all: clean-bins release
 else
 BAZEL_OPTS ?=
 BAZEL_BUILD_OPTS = --experimental_strict_action_env
-
-all: clean-bins envoy api
+all: clean-bins release
```
Comment (jrajahalme) [inline envoy/Makefile]: Why this change?

File: envoy/Makefile [hunk 3]
```diff
@@ -70,11 +102,7 @@ $(CHECK_FORMAT): force-non-root
 
 install: force-root
 	$(INSTALL) -m 0755 -d $(DESTDIR)$(BINDIR)
-	$(INSTALL) -m 0755 -T $(ENVOY_BIN) $(DESTDIR)$(BINDIR)/cilium-envoy
-# Strip only non-debug builds
-ifeq "$(findstring -dbg,$(realpath bazel-bin))" ""
-	$(STRIP) $(DESTDIR)$(BINDIR)/cilium-envoy
-endif
+	$(INSTALL) -m 0755 -T $(CILIUM_ENVOY_RELEASE_BIN) $(DESTDIR)$(BINDIR)/cilium-envoy
```
Comment (jrajahalme) [inline envoy/Makefile]: IMO we should keep the building with specific options and stripping as two separate steps, e.g., to allow unstripped release binaries to be run in the CI.

**Review by jrajahalme — APPROVED** [review]
Reviewing the build process only, as I don't have the required Istio knowledge to say much of the rest.

File: envoy/Dockerfile.istio_proxy_debug.in (hunk 1 above)
Comment (rlenglet) [reply envoy/Dockerfile.istio_proxy_debug.in:22]: The FROM is different.

File: envoy/Makefile (hunk 2 above)
Comment (rlenglet) [reply envoy/Makefile]: `release` already depended on `api`. This has the exact same effect, in the sense that the exact same targets are being depended on by `all`.

File: envoy/Makefile (hunk 3 above)
Comment (rlenglet) [reply envoy/Makefile]: OK, how about having other targets which produce non-stripped binaries? I would still like to keep stripping by default.

Comment (rlenglet) [reply envoy/Makefile]: Or better, explicitly don't strip for a few specific targets, e.g. `envoy-debug`. That would do what you want, right?

File: envoy/Dockerfile.istio_proxy_debug.in (hunk 1 above)
Comment (jrajahalme) [reply envoy/Dockerfile.istio_proxy_debug.in:22]: Oops, should not trust my eyes to make a diff :-)

File: envoy/Makefile (hunk 2 above)
Comment (jrajahalme) [reply envoy/Makefile]: You changed the build target, effectively from bazel "fastbuild" to "opt", you must have had a rationale for that?

File: envoy/Makefile (hunk 3 above)
Comment (jrajahalme) [reply envoy/Makefile]: The problem we have here it none of this matters if there is no problem. For the time being it would be really valuable to get good backtraces for any possible (if unlikely) segfaults, null dereferences etc. For that we need debug symbols and stack frame pointers. Running a non-optimized debug version would be even better, but it may be too slow. I have not experimented with that in the CI yet, one barrier being that a special box image with cached debug build would be needed. Also, when I did it locally, the unstripped binary was whopping 760MB :-)

Comment (jrajahalme) [reply envoy/Makefile]: I guess only stripping the release builds, and building that by default only when PKG_BUILD is passed in is a reasonable compromise. That way `make` will get you a non-release build (fastbuild) that still has the debug symbols. We should then make sure that at least the runtime tests in CI are run with that.

File: envoy/Makefile (hunk 2 above)
Comment (rlenglet) [reply envoy/Makefile]: Oh, I want to use `opt` for Istio to minimize the sidecar container size.
It doesn't matter as much for the node proxy used for dev. I'll update that.

Comment (jrajahalme) [reply envoy/Makefile]: `make` will build release (`opt` build) if PKG_BUILD is defined. docker image builds use that, maybe you could too, without changing the default make targets?

**Comment by rlenglet** [comment]
@jrajahalme I restored the build target for Cilium's Envoy binaries as they were. Please do another pass of review. Thanks!

**Comment by rlenglet** [comment]
test-me-please

File: envoy/Makefile [hunk 4]
```diff
@@ -128,7 +183,7 @@ endif
 endif
 
 force-non-root:
-ifeq ($(USER),root)
+ifneq ($(USER),root)
```
Comment (jrajahalme) [inline envoy/Makefile]: This flips the test while it should not.

**Review by jrajahalme — DISMISSED** [review]
`force-non-root` was changed to error out if the user is not root, while in intent is the opposite.

File: envoy/Makefile (hunk 4 above)
Comment (rlenglet) [reply envoy/Makefile]: Doh, left over from my manual testing.
The CI test failed because of that.
Sorry, will fix this.

**Comment by rlenglet** [comment]
test-me-please

**Comment by jrajahalme** [comment]
@rlenglet Check also the need to declare all phony targets individually as you did in this PR. AFAIK .PHONY is transitive, and if we make the force targets .PHONY, all targets depending on them will be phony as well. We need the binaries to be relinked at every build to get the latest version tag, so even binary targets should be phony, likely via a force target.

  Jarno

**Comment by rlenglet** [comment]
@jrajahalme that is fine, I checked that every call to `make` still links Envoy.

**Comment by rlenglet** [comment]
CI build failure:
https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-Validated/2865/execution/node/65/log/
```
19:45:17 • Failure in Spec Setup (BeforeEach) [440.296 seconds]
19:45:17 K8sValidatedChaosTest
19:45:17 /home/jenkins/workspace/Cilium-PR-Ginkgo-Tests-Validated/src/github.com/cilium/cilium/test/ginkgo-ext/scopes.go:280
19:45:17   Endpoint can still connect while Cilium is not running [BeforeEach]
19:45:17   /home/jenkins/workspace/Cilium-PR-Ginkgo-Tests-Validated/src/github.com/cilium/cilium/test/ginkgo-ext/scopes.go:312
19:45:17 
19:45:17   Pods are not ready after timeout
19:45:17   Expected
19:45:17       <*errors.errorString | 0xc421c74d80>: {
19:45:17           s: "Timeout reached: could not get Pods",
19:45:17       }
19:45:17   to be nil
19:45:17 
19:45:17   /home/jenkins/workspace/Cilium-PR-Ginkgo-Tests-Validated/src/github.com/cilium/cilium/test/k8sT/Chaos.go:61
```
I don't think this is related to this PR, which only touches `Makefile`.

**Comment by rlenglet** [comment]
test-me-please

**Comment by jrajahalme** [comment]
My point was that listing all phony targets explicitly easily gets out of sync over time and has little value.

**Comment by rlenglet** [comment]
Successful CI build: https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-Validated/2868/

**Comment by rlenglet** [comment]
@jrajahalme I removed the redundant phony targets. Thanks!

**Comment by rlenglet** [comment]
test-me-please

**Comment by rlenglet** [comment]
Unrelated build failure: https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-Validated/2911/

**Comment by rlenglet** [comment]
test-me-please

**Review by jrajahalme — APPROVED** [review]

**Comment by rlenglet** [comment]
test-me-please

**Comment by rlenglet** [comment]
test-me-please

**Comment by rlenglet** [comment]
test-me-please

**Comment by rlenglet** [comment]
Another unrelated CI build failure:
https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-Validated/2992/execution/node/65/log/
```
23:57:32 K8sValidatedPolicyTest
23:57:32 /home/jenkins/workspace/Cilium-PR-Ginkgo-Tests-Validated/src/github.com/cilium/cilium/test/ginkgo-ext/scopes.go:280
23:57:32   Basic Test [BeforeEach]
23:57:32   /home/jenkins/workspace/Cilium-PR-Ginkgo-Tests-Validated/src/github.com/cilium/cilium/test/ginkgo-ext/scopes.go:280
23:57:32     tests PolicyEnforcement updates
23:57:32     /home/jenkins/workspace/Cilium-PR-Ginkgo-Tests-Validated/src/github.com/cilium/cilium/test/ginkgo-ext/scopes.go:312
23:57:32 
23:57:32     Kubedns is not ready after timeout
23:57:32     Expected
23:57:32         <*errors.errorString | 0xc42036e450>: {
23:57:32             s: "Timeout reached: KubeDNS pods are not ready",
23:57:32         }
23:57:32     to be nil
23:57:32 
23:57:32     /home/jenkins/workspace/Cilium-PR-Ginkgo-Tests-Validated/src/github.com/cilium/cilium/test/k8sT/Policies.go:74
```
Known issue: https://github.com/cilium/cilium/issues/4014

**Comment by rlenglet** [comment]
test-me-please
